src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationBuilder.java

```diff
@@ -0,0 +1,194 @@
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+public class MutationBuilder {
+
+  private static final long MUTATION_LIMIT = 20000; // PLACEHOLDER
+  private static final long BATCH_LIMIT = 64000; // PLACEHOLDER
+
+  private boolean hasHeader;
+  private boolean isHeaderParsed;
+  private int mutationCount;
+  private int batchSize;
+  private int rowCount;
+  private List<Mutation> mutations;
+  private String tableName;
+  private Map<String, TypeCode> tableColumns;
+  private CSVFormat format;
+  private FileWriter fileWriter;
+
+  public MutationBuilder(
+      String tableName, Map<String, TypeCode> tableColumns, CSVFormat format, boolean hasHeader) {
+    this.mutationCount = 0;
+    this.batchSize = 0;
+    this.hasHeader = hasHeader;
+    this.isHeaderParsed = false;
+    this.tableName = tableName;
+    this.tableColumns = tableColumns;
+    this.format = format;
+    this.mutations = new ArrayList<>();
+  }
+
+  /** @return list of CopyData mutations not yet written into Spanner */
+  public List<Mutation> getMutations() {
+    return this.mutations;
+  }
+
+  /** @return number of rows copied into Spanner */
+  public int getRowCount() {
+    return this.rowCount;
+  }
+
+  /** Build mutation to add to mutations list with data contained within a CopyData payload */
+  public void buildMutation(ConnectionHandler connectionHandler, byte[] payload) throws Exception {
+    List<CSVRecord> records = parsePayloadData(payload);
+    if (!records.isEmpty()
+        && !payloadFitsInCurrentBatch(records.size() * records.get(0).size(), payload.length)) {
+      writeToSpanner(connectionHandler); // Close out the current batch and reset for the next batch
+    }
+    for (CSVRecord record : records) {
+      // Check that the number of columns in a record matches the number of columns in the table
+      if (record.size() != this.tableColumns.keySet().size()) {
+        writeToSpanner(connectionHandler);
+        throw new SQLException(
+            "Invalid COPY data: Row length mismatched. Expected "
+                + this.tableColumns.keySet().size()
+                + " columns, but only found "
+                + record.size());
+      }
+      WriteBuilder builder = Mutation.newInsertBuilder(this.tableName);
+      // Iterate through all table column to copy into
+      for (String columnName : this.tableColumns.keySet()) {
+        TypeCode columnType = this.tableColumns.get(columnName);
+        String recordValue = "";
+        try {
+          recordValue = record.get(columnName).trim();
+          switch (columnType) {
```

Comment:
Are we intentionally not supporting `NUMERIC` here?

Reply:
This code was originally written by an intern during the summer of 2021.  I think NUMERIC was not supported at that time.  We could add it now.